### ISSUE TITLE
[Question] How to run serial and parallel tests within single test file

### ISSUE THREAD

**Issue by aakash3771** [microsoft/playwright#16467]
I created a test file with 2 groups of tests using test.describe. For one group I set mode to serial and other to parallel. When I ran the test the tests in the test file, the parallel tests ran together but one serial test also started with them. My requirement is to run parallel tests in parallel, once they are finished then start the serial tests.

Also, I noted one thing in documentation that serial mode assumes the tests are dependent on each other which is not the case with me. What is the best way using features built in to Playwright?

**Comment by rwoll** [comment]
Thanks for the issue. I'd like to make sure I understand the scenario before answering. Given the following tests:

- A
- B
- C
- D
- E

Which ones can run at the same time as eachother? And which need to run one after the other, if any?

**Comment by aakash3771** [comment]
// feature 1.js
test.describe.parallel('parallel tests', () => {
  test('parallel one', async ({ page }) => {});

  test('parallel two', async ({ page }) => {});
});

test.describe.serial('serial tests', () => {
  test('serial one', async ({ page }) => {});

  test('serial two', async ({ page }) => {});
});

// feature 2.js
test.describe.parallel('parallel tests', () => {
  test('parallel one', async ({ page }) => {});

  test('parallel two', async ({ page }) => {});
});

test.describe.serial('serial tests', () => {
  test('serial one', async ({ page }) => {});

  test('serial two', async ({ page }) => {});
});

Requirements:
1. Run the parallel tests first, complete them. Once they are done, serial tests start. My serial tests have pre-condition and they are not inter-dependent.
2. I want a single Junit report for all above tests belonging to a feature. This will help in the import to a test management tool.

**Comment by rwoll** [comment]
Relates https://github.com/microsoft/playwright/pull/16698

**Comment by davidAtInleague** [comment]
is something like this possible:
```
test.describe.parallel(() => {
  test.describe.serial(() => {
    test(...) // A
    test(...) // B
  });
  test(...) // C
  test(...) // D
});
```
so B always follows A; but C can go before A or during A or after A or before D or after D, etc -- there's no ordering requirement on C or D.

**Comment by lsokolova** [comment]
Hello! 
We have a similar need:
- we have 3 test groups (test.describe) in my test file
- we need to run one of them first, and then 2 others can run in parallel.

Will the same [PR 16698](https://github.com/microsoft/playwright/pull/16698) be re-opened for the new API proposal implementation? I wonder if we can follow up on it or if there will be a new PR?

**Comment by lsokolova** [comment]
Hello @rwoll ! 
We have a similar need:
- we have 3 test groups (test.describe) in my test file
- we need to run one of them first, and then 2 others can run in parallel.

Will the same [PR 16698](https://github.com/microsoft/playwright/pull/16698) be re-opened for the new API proposal implementation? I wonder if we can follow up on it or if there will be a new PR?

**Comment by linustws** [comment]
I also need this feature. Currently I want to run tests batches in serial but tests within each batch in parallel. Whether it be running spec files in serial and tests within each file in parallel or spec files in serial, test groups in serial, test in each group in parallel, this feature would be very useful.

**Comment by Slapbox** [comment]
We test web and Android together and want to group our tests by feature-set, but without this we're forced to make separate test files for Android since they can only run in serial and never in parallel. Tests get wrongfully failed without serial mode enabled. Although perhaps our use case could be addressed in some other way.

**Comment by NathanielVill** [comment]
Hello @rwoll, 
I don't know if this is the right place but I believe I have a "Related" need. We have an application that needs to be tested in two languages (for now).
We need to run two different workers in a different language each, basically two workers testing in parallel the same spec.ts file, however, each spec.ts file has a list of tests that need to be executed in serial mode and finally, we need the spec.ts files to be tested also in the order they've been written (listed) inside Visual Studio Code. 
We have disabled parallelism via the config of Playwright but when we execute the tests we still find that Playwright executes the spec.ts in random order. We have also set inside the spec.ts files the describe.config to Serial mode but because the spec.ts files are not tested in the right order we get failed tests because they should have been tested before others were completed. 
Please find attached an image that we put together to try and better explain our need.
I am fairly new to playwright (only been using it for 6 months) so there are still a lot of things I don't know about the tool, however I am quite happy with it. 
I appreciate any help you can provide about this particular set up we need, and of course if any of what I explained doesn't make any sense, please let me know and I'll try to further elaborate. 
THank you so much.
![Screenshot 2023-09-21 at 15 41 55](https://github.com/microsoft/playwright/assets/122991496/4604fc95-ad74-4349-92b5-f92333a2feb2)

**Comment by Janis-BiCA-Acc** [comment]
whats the State at the moment of this, because I need this for our Tests too.

My Goal is to Achieve this:
![image](https://github.com/microsoft/playwright/assets/126558156/849f0b9d-737e-4424-a3a8-ec04639715b6)

I want to execute the first few tests serial that are creating something, and then checking with the parallel tests this creations 
At the moment I get a error that this isn't possible:
![image](https://github.com/microsoft/playwright/assets/126558156/a4cb429f-8554-443e-ab31-83374ce91d89)

**Comment by paradix86** [comment]
this feature in 2024 is a must in javascript/typescript automation framework.
We need the possibility to run, in serial test 1, and then in parallel tests 2, 3, 4, 5 6 etc...

Someone somehow with a workaround manages to solve this need with playwright in 2024?

**Comment by bdcarr** [comment]
Another use case for this kind of thing. I've got a test script for a data importer. Some imports rely on data from the previous imports being present in the system, but many of them can safely run parallel.

So I define three groups of test cases and want to run each group in order, but all the cases in a group can go in parallel. I hoped this would work, but it doesn't. It seems to cause every test case to run in parallel.

![image](https://github.com/microsoft/playwright/assets/1940488/56d7fbc5-b089-4349-b83b-250bb331b1fa)

**Comment by gabrielcartellivia** [comment]
![image](https://github.com/microsoft/playwright/assets/72199278/e6f0978b-59a5-4ec0-957f-0b095ff7d333)

https://playwright.dev/docs/test-projects#dependencies

I guess this feature solves most of the issues

**Comment by danielo515** [comment]
@gabrielcartellivia that does not solves the issue because they don't share context.

Each test will be completely isolated I want to be able to define variables in some of the serial tests that then I can use in the parallel ones.

**Comment by vBuravlev** [comment]
Playwrigth has a good message - do independent tests! But! There are entities in applications whose modification can affect the entire application. For example, the global pagination setting in the entire application. If it is changed in some test, then other tests may suffer because it is global. You have to create test dependencies one way or another. Now this is solved by launching projects separately, but it is not always convenient.

**Comment by omerchn** [comment]
Any updates about this?

**Comment by peterjp80** [comment]
Is there any solution for this yet?  I have a use case where I'm creating several different tasks for a task list with different settings.  These can all run in parallel. After they are all created, I need to test different view modes in the UI. Using dependencies with projects is inconvenient because these tests need to run on multiple browsers. Ideally, I would love to run a group of parallel tests within a group of serial tests.

`test.describe.serial('tasks', () => {
    test.describe.parallel('create tasks', () => {
        test('create a task without a label', async ({ page }) => {
        });

        test('create a task with a label', async ({ page }) => {
        });
    })

    test('view tasks in condensed mode', async ({ page }) => {
    });
});`

**Comment by alwaqfi** [comment]
I hope this feature will get implemented.

The use case I have is that one of the pages controls the visible and available menu to the users and at the same time will prevent accessing disabled pages through URL.

In my test suit, I can run the rest of the tests in parallel except the tests for this page which I need to run the last one otherwise some of the tests will fail.

The only thing I can do now is run everything in serial and leave the tests for that page to the end.
https://playwright.dev/docs/test-parallel#control-test-order
